File: app/src/main/java/com/android/nsystem/testapps/LoginPresenter.java
package com.android.nsystem.testapps;

public class LoginPresenter implements LoginContract.Presenter {

    private LoginContract.View mLoginView;

    LoginPresenter(LoginContract.View view) {
        this.mLoginView = view;
    }

    @Override
    public boolean isLoginInputValid(Account account) {
        if (account == null) {
            return false;
        } else {
            String username = account.getUsername();
            String password = account.getPassword();

            return username != null && !username.isEmpty() && password != null && !password.isEmpty();
        }
    }

    @Override
    public void submitLogin(Account account) {
        if (account != null) {
            if (isLoginInputValid(account)) {
                mLoginView.notifyLoginValid();
                mLoginView.clearLoginInput();
            }
        }
    }

    @Override
    public void loginInputTextChanged(Account account) {
        if (account == null) {
            mLoginView.disableLoginButton();
        } else {
            if (isLoginInputValid(account))
                mLoginView.enableLoginButton();
            else
                mLoginView.disableLoginButton();
        }
    }
}

File: app/src/main/java/com/android/nsystem/testapps/LoginPresenter.kt
package com.android.nsystem.testapps

import com.android.nsystem.testapps.utils.EspressoIdlingResource
import kotlinx.coroutines.*

/**
 * @author Putra Nugraha (putra.nugraha@dana.id)
 * @version LoginPresenter, v 0.1 13/12/20 23.37 by Putra Nugraha
 */
class LoginPresenter(private val view: LoginContract.View): LoginContract.Presenter {

    override fun isLoginInputValid(account: Account): Boolean = with(account) {
        return username.isNotBlank() && password.isNotBlank()
    }

    override fun submitLogin(account: Account) {
        EspressoIdlingResource.increment()
        CoroutineScope(Dispatchers.Main).launch {
            view.showProgress()
            delay(2000)
            EspressoIdlingResource.decrement()
            view.dismissProgress()
            if (isLoginInputValid(account)) {
                view.notifyLoginValid()
            }
        }
    }

    override fun loginInputTextChanged(account: Account) {
        when {
            isLoginInputValid(account) -> view.enableLoginButton()
            else -> view.disableLoginButton()
        }
    }
}